File: app/src/main/java/net/schueller/peertube/service/VideoPlayerService.java
/*
 * Copyright 2018 Stefan Schüller <sschueller@techdroid.com>
 *
 * License: GPL-3.0+
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package net.schueller.peertube.service;

import android.app.Notification;
import android.app.PendingIntent;
import android.app.PictureInPictureParams;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.media.AudioManager;
import android.net.Uri;
import android.os.Binder;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.webkit.URLUtil;
import androidx.annotation.Nullable;

import android.support.v4.media.MediaDescriptionCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.util.Log;
import android.util.Rational;
import android.widget.Toast;

import com.google.android.exoplayer2.C;
import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.ExoPlayerFactory;
import com.google.android.exoplayer2.PlaybackParameters;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.audio.AudioAttributes;
import com.google.android.exoplayer2.ext.mediasession.MediaSessionConnector;
import com.google.android.exoplayer2.ext.mediasession.TimelineQueueNavigator;
import com.google.android.exoplayer2.source.ExtractorMediaSource;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.ui.PlayerNotificationManager;
import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory;
import com.google.android.exoplayer2.util.Util;

import net.schueller.peertube.R;
import net.schueller.peertube.activity.VideoPlayActivity;
import net.schueller.peertube.helper.MetaDataHelper;
import net.schueller.peertube.model.Video;

import static android.media.session.PlaybackState.ACTION_PAUSE;
import static android.media.session.PlaybackState.ACTION_PLAY;
import static com.google.android.exoplayer2.ui.PlayerNotificationManager.ACTION_STOP;
import static net.schueller.peertube.activity.VideoListActivity.EXTRA_VIDEOID;

public class VideoPlayerService extends Service {

    private static final String TAG = "VideoPlayerService";

    private static final String MEDIA_SESSION_TAG = "peertube_player";

    private final IBinder mBinder = new LocalBinder();

    private static final String PLAYBACK_CHANNEL_ID = "playback_channel";

    private static final Integer PLAYBACK_NOTIFICATION_ID = 1;

    public SimpleExoPlayer player;

    private Video currentVideo;

    private String currentStreamUrl;

    private PlayerNotificationManager playerNotificationManager;

    private IntentFilter becomeNoisyIntentFilter = new IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY);

    private BecomingNoisyReceiver myNoisyAudioStreamReceiver = new BecomingNoisyReceiver();

    @Override
    public void onCreate() {
        Log.v(TAG, "onCreate...");

        super.onCreate();

        player = ExoPlayerFactory.newSimpleInstance(getApplicationContext(), new DefaultTrackSelector());

        // Stop player if audio device changes, e.g. headphones unplugged
        player.addListener(new Player.EventListener() {
            @Override
            public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {

                if (playbackState == ACTION_PAUSE) { // this means that pause is available, hence the audio is playing
                    Log.v(TAG, "ACTION_PLAY: " + playbackState);
                    registerReceiver(myNoisyAudioStreamReceiver, becomeNoisyIntentFilter);
                }

                if (playbackState
                        == ACTION_PLAY) { // this means that play is available, hence the audio is paused or stopped
                    Log.v(TAG, "ACTION_PAUSE: " + playbackState);
                    safeUnregisterReceiver();
                }
            }
        });

    }

    public class LocalBinder extends Binder {

        public VideoPlayerService getService() {
            // Return this instance of VideoPlayerService so clients can call public methods
            return VideoPlayerService.this;
        }
    }


    @Override
    public void onDestroy() {

        Log.v(TAG, "onDestroy...");
        if (playerNotificationManager != null) {
            playerNotificationManager.setPlayer(null);
        }
        //Was seeing an error when exiting the program about not unregistering the receiver.
        safeUnregisterReceiver();

        if (player != null) {
            player.release();
            player = null;
        }
        super.onDestroy();
    }

    private void safeUnregisterReceiver()
    {
        try {
            unregisterReceiver(myNoisyAudioStreamReceiver);
        } catch (Exception e) {
            Log.e("VideoPlayerService", "attempted to unregister a nonregistered service");
        }
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return mBinder;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Context context = this;
        Log.v(TAG, "onStartCommand...");

        if (!URLUtil.isValidUrl(currentStreamUrl)) {
            Toast.makeText(context, "Invalid URL provided. Unable to play video.", Toast.LENGTH_SHORT).show();
            return START_NOT_STICKY;
        } else {
            playVideo();
            return START_STICKY;
        }
    }


    public void setCurrentVideo(Video video) {
        Log.v(TAG, "setCurrentVideo...");
        currentVideo = video;
    }

    public void setCurrentStreamUrl(String streamUrl) {
        Log.v(TAG, "setCurrentStreamUrl..." + streamUrl);
        currentStreamUrl = streamUrl;
    }

    //Playback speed control
    public void setPlayBackSpeed(float speed) {
        Log.v(TAG, "setPlayBackSpeed...");
        player.setPlaybackParameters(new PlaybackParameters(speed));
    }

    /**
     * Returns the current playback speed of the player.
     *
     * @return the current playback speed of the player.
     */
    public float getPlayBackSpeed() {
        return player.getPlaybackParameters().speed;
    }

    public void playVideo() {
        Context context = this;

        // We need a valid URL

        Log.v(TAG, "playVideo...");

        // Produces DataSource instances through which media data is loaded.
        DataSource.Factory dataSourceFactory = new DefaultDataSourceFactory(getApplicationContext(),
                Util.getUserAgent(getApplicationContext(), "PeerTube"), null);

        // This is the MediaSource representing the media to be played.
        ExtractorMediaSource videoSource = new ExtractorMediaSource.Factory(dataSourceFactory)
                .createMediaSource(Uri.parse(currentStreamUrl));

        // Prepare the player with the source.
        player.prepare(videoSource);

        // Auto play
        player.setPlayWhenReady(true);

        //reset playback speed
        this.setPlayBackSpeed(1.0f);

        playerNotificationManager = PlayerNotificationManager.createWithNotificationChannel(
                context, PLAYBACK_CHANNEL_ID, R.string.playback_channel_name,
                PLAYBACK_NOTIFICATION_ID,
                new PlayerNotificationManager.MediaDescriptionAdapter() {

                    @Override
                    public String getCurrentContentTitle(Player player) {
                        return currentVideo.getName();
                    }

                    @Nullable
                    @Override
                    public PendingIntent createCurrentContentIntent(Player player) {
                        Intent intent = new Intent(context, VideoPlayActivity.class);
                        intent.putExtra(EXTRA_VIDEOID, currentVideo.getUuid());
                        return PendingIntent.getActivity(context, 0, intent,
                                PendingIntent.FLAG_UPDATE_CURRENT);
                    }

                    @Override
                    public String getCurrentContentText(Player player) {
                        return MetaDataHelper.getMetaString(
                                currentVideo.getCreatedAt(),
                                currentVideo.getViews(),
                                getBaseContext()
                        );
                    }

                    @Nullable
                    @Override
                    public Bitmap getCurrentLargeIcon(Player player,
                            PlayerNotificationManager.BitmapCallback callback) {
                        return null;
                    }
                }
        );

        playerNotificationManager.setSmallIcon(R.drawable.ic_logo_bw);

        // don't show skip buttons in notification
        playerNotificationManager.setUseNavigationActions(false);
        playerNotificationManager.setUseStopAction(true);

        playerNotificationManager.setNotificationListener(
                new PlayerNotificationManager.NotificationListener() {
                    @Override
                    public void onNotificationStarted(int notificationId, Notification notification) {
                        startForeground(notificationId, notification);
                    }

                    @Override
                    public void onNotificationCancelled(int notificationId) {
                        Log.v(TAG, "onNotificationCancelled...");
                        stopForeground(true);
                        Intent killFloat = new Intent(ACTION_STOP);
                        sendBroadcast(killFloat);
                   /*
                        Intent killFloat = new Intent(BROADCAST_ACTION);
                        Intent killFloatingWindow = new Intent(getApplicationContext(),VideoPlayActivity.class);
                        killFloatingWindow.putExtra("killFloat",true);

                        startActivity(killFloatingWindow);
                        // TODO: only kill the notification if we no longer have a bound activity
                        stopForeground(true);
                    */
                    }
                }
        );

        playerNotificationManager.setPlayer(player);

        // external Media control, Android Wear / Google Home etc.
        MediaSessionCompat mediaSession = new MediaSessionCompat(context, MEDIA_SESSION_TAG);
        mediaSession.setActive(true);
        playerNotificationManager.setMediaSessionToken(mediaSession.getSessionToken());
        MediaSessionConnector mediaSessionConnector = new MediaSessionConnector(mediaSession);
        mediaSessionConnector.setQueueNavigator(new TimelineQueueNavigator(mediaSession) {
            @Override
            public MediaDescriptionCompat getMediaDescription(Player player, int windowIndex) {
                return Video.getMediaDescription(context, currentVideo);
            }
        });
        mediaSessionConnector.setPlayer(player);

        // Audio Focus
        AudioAttributes audioAttributes = new AudioAttributes.Builder()
                .setUsage(C.USAGE_MEDIA)
                .setContentType(C.CONTENT_TYPE_MOVIE)
                .build();
        player.setAudioAttributes(audioAttributes, true);

    }

    // pause playback on audio output change
    private class BecomingNoisyReceiver extends BroadcastReceiver {

        @Override
        public void onReceive(Context context, Intent intent) {
            if (AudioManager.ACTION_AUDIO_BECOMING_NOISY.equals(intent.getAction())) {
                player.setPlayWhenReady(false);
            }
        }
    }


}

File: app/src/main/java/net/schueller/peertube/service/VideoPlayerService.kt
/*
 * Copyright (C) 2020 Stefan Schüller <sschueller@techdroid.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package net.schueller.peertube.service

import android.app.Notification
import net.schueller.peertube.helper.MetaDataHelper.getMetaString
import net.schueller.peertube.model.Video.Companion.getMediaDescription
import android.os.IBinder
import com.google.android.exoplayer2.ui.PlayerNotificationManager
import android.content.IntentFilter
import android.media.AudioManager
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector
import android.media.session.PlaybackState
import android.content.Intent
import android.widget.Toast
import net.schueller.peertube.helper.APIUrlHelper
import net.schueller.peertube.network.UnsafeOkHttpClient
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.hls.HlsMediaSource
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import com.google.android.exoplayer2.ui.PlayerNotificationManager.MediaDescriptionAdapter
import android.app.PendingIntent
import android.app.Service
import net.schueller.peertube.activity.VideoPlayActivity
import net.schueller.peertube.activity.VideoListActivity
import android.graphics.Bitmap
import android.support.v4.media.session.MediaSessionCompat
import com.google.android.exoplayer2.ext.mediasession.MediaSessionConnector
import com.google.android.exoplayer2.ext.mediasession.TimelineQueueNavigator
import android.support.v4.media.MediaDescriptionCompat
import android.content.BroadcastReceiver
import android.content.Context
import android.net.Uri
import android.os.Binder
import android.util.Log
import android.webkit.URLUtil
import androidx.core.app.NotificationCompat
import com.google.android.exoplayer2.*
import com.google.android.exoplayer2.audio.AudioAttributes
import com.google.android.exoplayer2.ext.okhttp.OkHttpDataSource
import com.google.android.exoplayer2.ui.PlayerNotificationManager.NotificationListener
import net.schueller.peertube.R.drawable
import net.schueller.peertube.R.string
import net.schueller.peertube.model.Video
import java.lang.Exception

class VideoPlayerService : Service() {

    private val mBinder: IBinder = LocalBinder()
    @JvmField
    var player: ExoPlayer? = null
    private var currentVideo: Video? = null
    private var currentStreamUrl: String? = null
    private var currentStreamUrlIsHLS = false
    private var playerNotificationManager: PlayerNotificationManager? = null
    private val becomeNoisyIntentFilter = IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY)
    private val myNoisyAudioStreamReceiver = BecomingNoisyReceiver()
    override fun onCreate() {
        Log.v(TAG, "onCreate...")
        super.onCreate()
        player = ExoPlayer.Builder(applicationContext)
            .setTrackSelector(DefaultTrackSelector(applicationContext))
            .build()

        // Stop player if audio device changes, e.g. headphones unplugged
        player!!.addListener(object : Player.Listener {
            override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
                if (playbackState.toLong() == PlaybackState.ACTION_PAUSE) { // this means that pause is available, hence the audio is playing
                    Log.v(TAG, "ACTION_PLAY: $playbackState")
                    registerReceiver(myNoisyAudioStreamReceiver, becomeNoisyIntentFilter)
                }
                if (playbackState
                        .toLong() == PlaybackState.ACTION_PLAY
                ) { // this means that play is available, hence the audio is paused or stopped
                    Log.v(TAG, "ACTION_PAUSE: $playbackState")
                    safeUnregisterReceiver()
                }
            }
        })
    }

    inner class LocalBinder : Binder() {

        // Return this instance of VideoPlayerService so clients can call public methods
        val service: VideoPlayerService
            get() =// Return this instance of VideoPlayerService so clients can call public methods
                this@VideoPlayerService
    }

    override fun onDestroy() {
        Log.v(TAG, "onDestroy...")
        if (playerNotificationManager != null) {
            playerNotificationManager!!.setPlayer(null)
        }
        //Was seeing an error when exiting the program about not unregistering the receiver.
        safeUnregisterReceiver()
        if (player != null) {
            player!!.release()
            player = null
        }
        super.onDestroy()
    }

    private fun safeUnregisterReceiver() {
        try {
            unregisterReceiver(myNoisyAudioStreamReceiver)
        } catch (e: Exception) {
            Log.e("VideoPlayerService", "attempted to unregister a non-registered service")
        }
    }

    override fun onBind(intent: Intent): IBinder {
        return mBinder
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        val context: Context = this
        Log.v(TAG, "onStartCommand...")
        return if (!URLUtil.isValidUrl(currentStreamUrl)) {
            Toast.makeText(context, "Invalid URL provided. Unable to play video.", Toast.LENGTH_SHORT).show()
            START_NOT_STICKY
        } else {
            playVideo()
            START_STICKY
        }
    }

    fun setCurrentVideo(video: Video?) {
        Log.v(TAG, "setCurrentVideo...")
        currentVideo = video
    }

    fun setCurrentStreamUrl(streamUrl: String, isHLS: Boolean) {
        Log.v(TAG, "setCurrentStreamUrl...$streamUrl")
        currentStreamUrlIsHLS = isHLS
        currentStreamUrl = streamUrl
    }

    /**
     * Returns the current playback speed of the player.
     *
     * @return the current playback speed of the player.
     *///Playback speed control
    var playBackSpeed: Float
        get() = player!!.playbackParameters.speed
        set(speed) {
            Log.v(TAG, "setPlayBackSpeed...")
            player!!.playbackParameters = PlaybackParameters(speed)
        }

    private fun playVideo() {
        val context: Context = this

        // We need a valid URL
        Log.v(TAG, "playVideo...")

        // Produces DataSource instances through which media data is loaded.
        val okhttpClientBuilder: okhttp3.OkHttpClient.Builder = if (!APIUrlHelper.useInsecureConnection(this)) {
            okhttp3.OkHttpClient.Builder()
        } else {
            UnsafeOkHttpClient.getUnsafeOkHttpClientBuilder()
        }

        // Create a data source factory.
        val dataSourceFactory:  OkHttpDataSource.Factory =  OkHttpDataSource.Factory(
            okhttpClientBuilder.build()
        )

        // Create a progressive media source pointing to a stream uri.
        val mediaSource: MediaSource = if (currentStreamUrlIsHLS) {
            HlsMediaSource.Factory(dataSourceFactory)
                .createMediaSource(MediaItem.fromUri(Uri.parse(currentStreamUrl)))
        } else {
            ProgressiveMediaSource.Factory(dataSourceFactory)
                .createMediaSource(MediaItem.fromUri(Uri.parse(currentStreamUrl)))
        }

        // Set the media source to be played.
        player!!.setMediaSource(mediaSource)

        // Prepare the player.
        player!!.prepare()

        // Auto play
        player!!.playWhenReady = true

        //set playback speed to global default
        val sharedPref = getSharedPreferences(
            packageName + "_preferences",
            Context.MODE_PRIVATE
        )

        val speed = sharedPref.getString(getString(string.pref_video_speed_key), "1.0")!!.toFloat()
        playBackSpeed = speed

        playerNotificationManager = PlayerNotificationManager.Builder(
            this,
            PLAYBACK_NOTIFICATION_ID,
            PLAYBACK_CHANNEL_ID,
        ).setMediaDescriptionAdapter(
                object : MediaDescriptionAdapter {
                    override fun getCurrentContentTitle(player: Player): CharSequence {
                        return currentVideo!!.name
                    }

                    override fun createCurrentContentIntent(player: Player): PendingIntent? {
                        val intent = Intent(context, VideoPlayActivity::class.java)
                        intent.putExtra(VideoListActivity.EXTRA_VIDEOID, currentVideo!!.uuid)
                        return PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT)

                    }

                    override fun getCurrentContentText(player: Player): CharSequence {
                        return getMetaString(
                                currentVideo!!.createdAt,
                                currentVideo!!.views,
                                baseContext
                        )
                    }

                    override fun getCurrentSubText(player: Player): CharSequence { return ""}

                    override fun getCurrentLargeIcon(
                            player: Player,
                            callback: PlayerNotificationManager.BitmapCallback
                    ): Bitmap? {
                        return null
                    }
                }
        ).setNotificationListener(
            object : NotificationListener {

                override fun onNotificationPosted(notificationId: Int, notification: Notification, ongoing: Boolean) {
                    super.onNotificationPosted(notificationId, notification, ongoing)
                    if (ongoing) // allow notification to be dismissed if player is stopped
                        startForeground(notificationId, notification)
                    else
                        stopForeground(false)
                }

                override fun onNotificationCancelled(notificationId: Int, dismissedByUser: Boolean) {
                    super.onNotificationCancelled(notificationId, dismissedByUser)
                    stopSelf()
                    stopForeground(true)
                }
            }
        ).setChannelNameResourceId(string.playback_channel_name)
        .setChannelDescriptionResourceId(string.playback_channel_description)
        .build()

        playerNotificationManager!!.setPriority(NotificationCompat.PRIORITY_DEFAULT)

        playerNotificationManager!!.setSmallIcon(drawable.ic_logo_bw)

        // don't show skip buttons in notification
        playerNotificationManager!!.setUseNextAction(false)
        playerNotificationManager!!.setUsePreviousAction(false)

        playerNotificationManager!!.setPlayer(player)

        // external Media control, Android Wear / Google Home etc.
        val mediaSession = MediaSessionCompat(context, MEDIA_SESSION_TAG)
        mediaSession.isActive = true
        playerNotificationManager!!.setMediaSessionToken(mediaSession.sessionToken)
        val mediaSessionConnector = MediaSessionConnector(mediaSession)
        mediaSessionConnector.setQueueNavigator(object : TimelineQueueNavigator(mediaSession) {
            override fun getMediaDescription(player: Player, windowIndex: Int): MediaDescriptionCompat {
                return getMediaDescription(currentVideo!!)
            }
        })
        mediaSessionConnector.setPlayer(player)

        // Audio Focus
        val audioAttributes = AudioAttributes.Builder()
            .setUsage(C.USAGE_MEDIA)
            .setContentType(C.CONTENT_TYPE_MOVIE)
            .build()
        player!!.setAudioAttributes(audioAttributes, true)
    }

    // pause playback on audio output change
    private inner class BecomingNoisyReceiver : BroadcastReceiver() {

        override fun onReceive(context: Context, intent: Intent) {
            if (AudioManager.ACTION_AUDIO_BECOMING_NOISY == intent.action) {
                player!!.playWhenReady = false
            }
        }
    }

    companion object {

        private const val TAG = "VideoPlayerService"
        private const val MEDIA_SESSION_TAG = "peertube_player"
        private const val PLAYBACK_CHANNEL_ID = "playback_channel"
        private const val PLAYBACK_NOTIFICATION_ID = 1
    }
}